segmentstore/storage/impl/src/main/java/io/pravega/segmentstore/storage/impl/hdfs/HDFSStorage.java

```diff
@@ -196,86 +328,302 @@ public boolean supportsTruncation() {
+        } catch (IOException e) {
+            throw HDFSExceptionHelpers.throwException(streamSegmentName, e);
+        }
+        if (existingFiles != null && existingFiles.length > 0) {
+            // Segment already exists; don't bother with anything else.
+            throw HDFSExceptionHelpers.throwException(streamSegmentName, HDFSExceptionHelpers.segmentExistsException(streamSegmentName));
+        }
 
+        // Create the first file in the segment.
+        Path fullPath = getFilePath(streamSegmentName, this.epoch);
+        try {
+            // Create the file, and then immediately close the returned OutputStream, so that HDFS may properly create the file.
+            this.fileSystem.create(fullPath, READWRITE_PERMISSION, false, 0, this.config.getReplication(),
+                    this.config.getBlockSize(), null).close();
+            log.debug("Created '{}'.", fullPath);
+        } catch (IOException e) {
+            throw HDFSExceptionHelpers.throwException(streamSegmentName, e);
+        }
+        LoggerHelpers.traceLeave(log, "create", traceId, streamSegmentName);
+        return StreamSegmentInformation.builder().name(streamSegmentName).build();
+    }
     //endregion
 
     //region Helpers
 
     /**
-     * Executes the given Callable and returns its result, while translating any Exceptions bubbling out of it into
-     * StreamSegmentExceptions.
-     *
-     * @param operation The function to execute.
+     * Casts the given handle as a HDFSSegmentHandle that has isReadOnly == false.
      */
-    private <T extends FileSystemOperation & RunnableWithException> void run(T operation) throws StreamSegmentException {
-        ensureInitializedAndNotClosed();
-        try {
-            operation.run();
-        } catch (Exception e) {
-            throwException(e, operation);
+    private HDFSSegmentHandle asWritableHandle(SegmentHandle handle) {
+        Preconditions.checkArgument(!handle.isReadOnly(), "handle must not be read-only.");
+        return asReadableHandle(handle);
+    }
+
+    /**
+     * Casts the given handle as a HDFSSegmentHandle irrespective of its isReadOnly value.
+     */
+    private HDFSSegmentHandle asReadableHandle(SegmentHandle handle) {
+        Preconditions.checkArgument(handle instanceof HDFSSegmentHandle, "handle must be of type HDFSSegmentHandle.");
+        return (HDFSSegmentHandle) handle;
+    }
+
+    private void ensureInitializedAndNotClosed() {
+        Exceptions.checkNotClosed(this.closed.get(), this);
+        Preconditions.checkState(this.fileSystem != null, "HDFSStorage is not initialized.");
+    }
+
+    //endregion
+
+    //Region HDFS helper methods.
+
+    /**
+     * Gets an array (not necessarily ordered) of FileStatus objects currently available for the given Segment.
+     * These must be in the format specified by NAME_FORMAT (see EXAMPLE_NAME_FORMAT).
+     */
+    FileStatus[] findAllRaw(String segmentName) throws IOException {
```

Comment:
Since this always returns 0 or 1 files, please change this so it doesn't return an array. Also fix the javadoc